pkg/solana/logpoller/log_poller.go

```diff
@@ -510,3 +513,32 @@ func (lp *Service) backgroundWorkerRun(ctx context.Context) {
 func (lp *Service) FilteredLogs(ctx context.Context, queryFilter []query.Expression, limitAndSort query.LimitAndSort, queryName string) ([]Log, error) {
 	return lp.orm.FilteredLogs(ctx, queryFilter, limitAndSort, queryName)
 }
+
+func (lp *Service) computeLookbackWindow(ctx context.Context) (int64, error) {
+	latestFinalizedSlot, err := lp.client.SlotHeightWithCommitment(ctx, rpc.CommitmentFinalized)
+	if err != nil {
+		return 0, fmt.Errorf("error getting latest slot from RPC: %w", err)
+	}
+
+	if latestFinalizedSlot == 0 {
+		return 0, fmt.Errorf("latest finalized slot is 0 - waiting for next slot to start processing")
+	}
+
+	lookback := lp.filters.MaxRetention()
+	if lookback == 0 { // Use default lookback if all filters have permanent retention or less than
+		lookback = DefaultLookbackWindow
+	}
+
+	firstAvailableSlot, err := lp.client.GetFirstAvailableBlock(ctx) // Despite the name, this returns slot # not block number
+
+	// nolint:gosec
+	// G115: integer overflow conversion uint64 -&gt; int64
+	latestSlot, firstSlot := int64(latestFinalizedSlot), int64(firstAvailableSlot)
+	lookbackSlot := latestSlot - int64((lookback-1)/lp.blockTime) - 1 // = latestSlot - ceil(lookback/blockTime)
+
+	// This is an optimization to avoid requesting pruned blocks. If there's an err we can just ignore
+	if err == nil && firstSlot > lookbackSlot {
```

Comment:
nit (feel free to ignore): it would nice to move call get get GetFirstAvailableBlock closer to the actual error handling

Reply:
Moved!  (My original motivation was to group all of the integer conversion together so I didn't need two different lint exclusions, but I agree it ended up making the logic a lot more confusing to read.)